File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement{
    private MapDirection direction;
    private Vector2d position;

    public Animal(){
        this.direction = MapDirection.NORTH;
        this.position = new Vector2d(2,2);
    }

    public Animal(MapDirection direction, Vector2d positions){
        this.direction = direction;
        this.position = positions;
    }

    @Override
    public Vector2d getPosition() {
        return this.position;
    }

    public MapDirection getDirection() {
        return this.direction;
    }

    public String toString(){
        return this.direction.toString();
    }

    public boolean isAt(Vector2d position){
        return this.position.equals(position);
    }

    public void move(MoveDirection direction, WorldMap map){
        Vector2d potentialNewPosition;

        switch (direction){
            case RIGHT:
                this.direction = this.direction.next();
                break;
            case LEFT:
                this.direction = this.direction.previous();
                break;
            case FORWARD:
                potentialNewPosition = this.position.add(this.direction.toUnitVector());
                if (map.canMoveTo(potentialNewPosition)) {
                    if (map.objectAt(potentialNewPosition) instanceof Grass) {
                        ((GrassField) map).RespawnGrass(potentialNewPosition);
                    }
                    this.position = potentialNewPosition;
                }
                break;
            case BACKWARD:
                potentialNewPosition = this.position.subtract(this.direction.toUnitVector());
                if (map.canMoveTo(potentialNewPosition)) {
                    if (map.objectAt(potentialNewPosition) instanceof Grass) {
                        ((GrassField) map).RespawnGrass(potentialNewPosition);
                    }
                    this.position = potentialNewPosition;
                }
                break;
            default:
                break;
        }
    }
}

File: proj-lab2/src/main/kotlin/model/Animal.kt
package model

import utils.MapDirection
import utils.MoveDirection
import utils.Vector2d
import utils.toUnitVector

class Animal (var position: Vector2d, var direction: MapDirection, private var map: IWorldMap) {
    fun isAt(positionToCheck: Vector2d?): Boolean {
        return position == positionToCheck
    }

    fun move(direction: MoveDirection) {
        when (direction) {
            MoveDirection.RIGHT -> this.direction = this.direction.next()
            MoveDirection.LEFT -> this.direction = this.direction.previous()
            else -> {
                var directionVector: Vector2d = this.direction.toUnitVector()
                if (direction === MoveDirection.BACKWARD) {
                    directionVector = directionVector.opposite()
                }
                val newLocation: Vector2d = position.add(directionVector)
                if (map.canMoveTo(newLocation)) {
                    position = newLocation
                }
            }
        }
    }

    override fun toString(): String = when (direction) {
        MapDirection.EAST -> "E"
        MapDirection.WEST -> "W"
        MapDirection.NORTH -> "N"
        MapDirection.SOUTH -> "S"
        else -> ""
    }
}